File: cortanavcds/cs/adventureworks/viewmodel/triplistviewmodel.cs
//*********************************************************
//
// Copyright (c) Microsoft. All rights reserved.
// This code is licensed under the MIT License (MIT).
// THIS CODE IS PROVIDED *AS IS* WITHOUT WARRANTY OF
// ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING ANY
// IMPLIED WARRANTIES OF FITNESS FOR A PARTICULAR
// PURPOSE, MERCHANTABILITY, OR NON-INFRINGEMENT.
//
//*********************************************************

using AdventureWorks.Common;
using AdventureWorks.Model;
using AdventureWorks.View;
using System;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Windows.Input;

namespace AdventureWorks.ViewModel
{
    /// <summary>
    /// View Model controlling the behavior of a List view of trips 
    /// </summary>
    public class TripListViewModel : ViewModelBase
    {
        private ICommand addTripCommand;
        private ObservableCollection<Trip> trips;
        private Trip selectedTrip;
        private TripStore store;

        /// <summary>
        /// Construct the trip view, passing in the persistent trip store. Sets up
        /// a command to handle invoking the Add button.
        /// </summary>
        /// <param name="store"></param>
        public TripListViewModel(TripStore store)
        {
            this.store = store;
            Trips = store.Trips;

            addTripCommand = new RelayCommand(new Action(AddTrip));
        }

        /// <summary>
        /// The list of trips to display on the UI.
        /// </summary>
        public ObservableCollection<Trip> Trips
        {
            get
            {
                return trips;
            }
            private set
            {
                trips = value;
                NotifyPropertyChanged("Trips");
            }
        }

        /// <summary>
        /// A two-way binding that keeps reference to the currently selected trip on
        /// the UI.
        /// </summary>
        public Trip SelectedTrip
        {
            get
            {
                return selectedTrip;
            }
            set
            {
                selectedTrip = value;
                NotifyPropertyChanged("SelectedTrip");
            }
        }

        /// <summary>
        /// The command to invoke when the Add button is pressed.
        /// </summary>
        public ICommand AddTripCommand
        {
            get
            {
                return addTripCommand;
            }
        }

        /// <summary>
        /// Reload the trip store from data files.
        /// </summary>
        internal async Task LoadTrips()
        {
            await store.LoadTrips();
        }


        /// <summary>
        /// The implementation of the command to execute when the Add button is pressed.
        /// </summary>
        private void AddTrip()
        {
            App.NavigationService.Navigate<TripDetails>();
        }

        /// <summary>
        /// Handles a user selecting a trip on the UI by navigating to the TripDetails view
        /// for that trip.
        /// </summary>
        internal void SelectionChanged()
        {
            if (SelectedTrip != null)
            {
                App.NavigationService.Navigate<TripDetails>(SelectedTrip);
            }
        }
    }
}

File: Samples/CortanaVoiceCommand/cpp/AdventureWorks/ViewModel/TripListViewModel.cpp
//*********************************************************
//
// Copyright (c) Microsoft. All rights reserved.
// This code is licensed under the MIT License (MIT).
// THIS CODE IS PROVIDED *AS IS* WITHOUT WARRANTY OF
// ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING ANY
// IMPLIED WARRANTIES OF FITNESS FOR A PARTICULAR
// PURPOSE, MERCHANTABILITY, OR NON-INFRINGEMENT.
//
//*********************************************************
#include "pch.h"
#include "ViewModel\TripListViewModel.h"

#include "Common\RelayCommand.h"
#include "View\TripDetails.xaml.h"

using namespace AdventureWorks;

using namespace AdventureWorks_Shared;
using namespace Concurrency;
using namespace Platform;
using namespace Windows::ApplicationModel::Core;
using namespace Windows::Foundation;
using namespace Windows::Foundation::Collections;
using namespace Windows::UI::Xaml;
using namespace Windows::UI::Xaml::Controls;
using namespace Windows::UI::Xaml::Input;

/// <summary>
/// Sets up list view and command binding references.
/// </summary>
/// <param name="store">Shared reference to trip store.</param>
TripListViewModel::TripListViewModel(AdventureWorks_Shared::TripStore ^ store)
{
    this->store = store;
    this->trips = store->Trips;

    addTripCommand = ref new RelayCommand(
        [this](Object^ parameter) {return true; },
        [this](Object^ parameter) {this->AddTrip(); });
}

/// <summary> Command executed when new trip button is clicked </summary>
void TripListViewModel::AddTrip()
{
    auto rootFrame = dynamic_cast<Frame^>(Window::Current->Content);
    rootFrame->Navigate(TripDetails::typeid);
}

Trip^ TripListViewModel::SelectedTrip::get()
{
    return this->selectedTrip;
}

void TripListViewModel::SelectedTrip::set(Trip^ value)
{
    this->selectedTrip = value;
    NotifyPropertyChanged("SelectedTrip");
}

ICommand^ TripListViewModel::AddTripCommand::get()
{
    return this->addTripCommand;
}

/// <summary> Triggers asynchronous initialization of trip store assets </summary>
void TripListViewModel::LoadTrips()
{
    store->LoadTrips();
}

/// <summary> Handle clicking on a different trip to navigate to that trip </summary>
void TripListViewModel::SelectionChanged()
{
    if (this->SelectedTrip != nullptr)
    {
        auto rootFrame = dynamic_cast<Frame^>(Window::Current->Content);
        rootFrame->Navigate(TripDetails::typeid, this->SelectedTrip);
    }
}

IObservableVector<AdventureWorks_Shared::Trip^>^ TripListViewModel::Trips::get()
{
    return this->trips;
}
